### DISCUSSION TITLE
No Event are being Listened When subscribed, version 1.10

### DISCUSSION THREAD

**Discussion by vinaye42** [imdhemy/laravel-in-app-purchases#408] [General]
`
website.test.com:443 172.70.43.153 - - [15/May/2024:06:43:26 +0000] "POST /liap/notifications?signature=<signature>&provider=app-store HTTP/1.1" 200 4043 "-" "Apache-HttpClient/4.5.14 (Java/11.0.23)"
'routing' => [
        'signed' => true,
],
'eventListeners' => [
        Subscribed::class => [ \App\Listeners\BuySubscription::class ],
],

namespace App\Listeners;

use Imdhemy\Purchases\Events\AppStore\Subscribed;
class BuySubscription
{

    public function handle(Subscribed $event)
    {
        info('subscribed-event');
        info($event);
    }
}
but still didn't listened event, 
we follow the document added all the detail as mention in document, should it take time 
`

**Comment by imdhemy** [comment]
Did you try other events? 

Regrading the Subscribed event you follow this discussion. #24

**Reply by vinaye42** [reply]
Hi @imdhemy,
No we haven't checked other events.
We have also tried with out signature, 
'routing' => [
'signed' => false,
],

but still didn't get log in above BuySubscription Listeners

**Comment by vinaye42** [comment]
hi @imdhemy,
running the below artisan command,
php artisan liap:apple:test-notification --sandbox
We get Test notification token in terminal,
But we didn't get log in **AppStoreV2NotificationHandler.php**
`
if ($serverNotification->isTest()) {
            Log::info(
                'AppStoreV2NotificationHandler: Test notification received ' .
                $this->request->get('signedPayload')
            );
            return;
`
also in our listeners handler.

**Comment by vinaye42** [comment]
Hi @imdhemy,
![storeV2Notification](https://github.com/imdhemy/laravel-in-app-purchases/assets/93908879/8ca93b18-8f62-4750-8288-ade4015ddd4c)
We got the below log in laravel,
[2024-05-21 10:08:47] production.INFO: store v2 test  
[2024-05-21 10:08:47] production.INFO: store v2 handle

**Comment by vinaye42** [comment]
Hi @imdhemy,
php artisan liap:apple:test-notification --sandbox
We get Test notification token in terminal,
![liap](https://github.com/imdhemy/laravel-in-app-purchases/assets/93908879/97c8fbdb-c41d-4cd9-a2e7-cd1c0483a8bc)

here we didn't get that log in BuySubscription.php

**Comment by imdhemy** [comment]
Hi @vinaye42 
The LIAP test command `php artisan liap:apple:test-notification --sandbox`  is not meant to test the subscription life cycle.  It is supposed to check if your configuration works as expected. 

| Event | Meaning |
| --- | --- |
| Received the notification token | AppStore received your request and replied with a token |
| The Log of the test notification present | Your setup with the AppStore works as expected & your app can receive other notifications |

That means the test notification should not trigger the [AppStore events](https://github.com/imdhemy/laravel-in-app-purchases/tree/cc4f7978abeea5e401b46d63e587772fbe598fe5/src/Events/AppStore). 

For more information on how to test the subscription life-cycle, you can consult [this guide from Apple](https://developer.apple.com/documentation/storekit/in-app_purchase/testing_in-app_purchases_with_sandbox). 

To know how to handle events in your Laravel application, you can consult the LIAP documentation.

- [Event Listeners](https://imdhemy.com/laravel-iap-docs/docs/get-started/event-listeners)
- [Available Events.](https://imdhemy.com/laravel-iap-docs/docs/server-notifications/event-list)

It would help if you double-check which version of notifications you are using.

**Comment by vinaye42** [comment]
Hi @imdhemy,
Thanks for the response.
We are using version 2 notification.
We get the solution, we just change in eventListeners in config/liap.php as below,
> 
`'eventListeners' => [`
`    \Imdhemy\Purchases\Events\AppStore\Subscribed::class => [`
`          \App\Listeners\BuySubscription::class,`
 `    ],`
` ],`

 and we get the log in BuySubscription.php listeners, but now we added below line
 
`$subscription = $event->getServerNotification()->getSubscription();`
`$uniqueIdentifier = $subscription->getUniqueIdentifier();`
`$expirationTime = $subscription->getExpiryTime();`

that's gives error,
`production.ERROR: Typed property Imdhemy\Purchases\Events\PurchaseEvent::$serverNotification must not be accessed before initialization {"exception":"[object] (Error(code: 0): Typed property Imdhemy\\Purchases\\Events\\PurchaseEvent::$serverNotification must not be accessed before initialization at /var/www/html/client/vendor/imdhemy/laravel-purchases/src/Events/PurchaseEvent.php:33)
[stacktrace]
#0 /var/www/html/client/app/Listeners/BuySubscription.php(25): Imdhemy\\Purchases\\Events\\PurchaseEvent->getServerNotification()
`

**Reply by vinaye42** [reply]
Hi @imdhemy 
can you please verify this, we are stuck at here

**Reply by imdhemy** [reply]
I will make time for it.

**Reply by imdhemy** [reply]
Hey @vinaye42 

- How can I reproduce it?
- Which version of App store notifications you are using?
- Can you provide the payload you received from your sandbox server notification?

**Reply by vinaye42** [reply]
Hello @imdhemy,
We have fixed this using the below code, check the code and let us know if anything is wrong

`<?php`

`namespace App\Listeners;`

`use App\Models\User;`
`use Imdhemy\Purchases\Events\AppStore\Subscribed;`
`use Illuminate\Support\Facades\Log;`

`class BuySubscription`
`{`
`    public function handle(Subscribed $event)`
`    {`
`        info('subscribed-event');`
`        // Check if serverNotification is initialized`
`       if (is_null($event->getServerNotification())) {`
`            Log::info('ServerNotification is not initialized.');`
`            return;`
`       }`
`        $notification = $event->getServerNotification();`
`        info('after notification - event');`

`        if (is_null($notification->getSubscription())) {`
`            Log::info('Subscription is not initialized.');`
`            return;`
`        }`
`        $subscription = $notification->getSubscription();`
`        $uniqueIdentifier = $subscription->getUniqueIdentifier();`
`        $expirationTime = $subscription->getExpiryTime();`
`       $provider = $subscription->getProvider();`
`    }`
`}`
